File: PL-Benchmark/algorithms/binary_search/implementations/python/src/binary_search.py
#!/usr/bin/env python3

import sys
import time

def binary_search(arr, target):
    left, right = 0, len(arr) - 1

    while left <= right:
        mid = (left + right) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            left = mid + 1
        else:
            right = mid - 1

    return -1  # Target not found

def main():
    if len(sys.argv) < 3:
        print("Usage: python binary_search.py <input_file> <targets_file>")
        sys.exit(1)

    input_file = sys.argv[1]
    targets_file = sys.argv[2]

    try:
        # Read sorted array
        with open(input_file, 'r') as f:
            numbers = [int(line.strip()) for line in f if line.strip()]

        # Read targets
        with open(targets_file, 'r') as f:
            targets = [int(line.strip()) for line in f if line.strip()]

        # Search for each target
        found = 0
        not_found = 0
        results = []

        for target in targets:
            index = binary_search(numbers, target)
            if index != -1:
                found += 1
                results.append(f"{target}: Found at index {index}")
            else:
                not_found += 1
                results.append(f"{target}: Not found")

        # Write results
        with open('binary_search_results.txt', 'w') as f:
            for result in results:
                f.write(f"{result}\n")

        print(f"Search results: found {found}, not found {not_found}")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: PL-Benchmark/algorithms/binary_search/implementations/typescript/src/binary_search.ts
/**
 * Binary Search implementation in TypeScript
 *
 * Usage: node binary_search.js <input_file_path> <targets_file_path>
 */

import * as fs from 'fs';

/**
 * Binary search algorithm implementation
 * @param arr Sorted array to search in
 * @param target Value to search for
 * @returns Index of the target if found, -1 otherwise
 */
function binarySearch(arr: number[], target: number): number {
    let left = 0;
    let right = arr.length - 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);

        if (arr[mid] === target) {
            return mid; // Found the target
        } else if (arr[mid] < target) {
            left = mid + 1; // Search in the right half
        } else {
            right = mid - 1; // Search in the left half
        }
    }

    return -1; // Target not found
}

/**
 * Main function to handle file I/O and searching
 */
function main(): void {
    // Check command line arguments
    if (process.argv.length < 4) {
        console.error('Usage: node binary_search.js <input_file_path> <targets_file_path>');
        process.exit(1);
    }

    const inputFile = process.argv[2];
    const targetsFile = process.argv[3];

    try {
        // Read and parse the sorted input array
        const inputData = fs.readFileSync(inputFile, 'utf8');
        const sortedArray = inputData.trim().split('\n').map(line => {
            const num = parseInt(line.trim(), 10);
            if (isNaN(num)) {
                throw new Error(`Invalid number in input file: ${line}`);
            }
            return num;
        });

        // Read and parse the targets to search for
        const targetsData = fs.readFileSync(targetsFile, 'utf8');
        const targets = targetsData.trim().split('\n').map(line => {
            const num = parseInt(line.trim(), 10);
            if (isNaN(num)) {
                throw new Error(`Invalid number in targets file: ${line}`);
            }
            return num;
        });

        // Search for each target and count found/not found
        let found = 0;
        let notFound = 0;

        const results: Array<{target: number, index: number, found: boolean}> = [];

        for (const target of targets) {
            const index = binarySearch(sortedArray, target);
            const isFound = index !== -1;

            results.push({
                target,
                index,
                found: isFound
            });

            if (isFound) {
                found++;
            } else {
                notFound++;
            }
        }

        // Write results summary
        console.log(`Search results: found ${found}, not found ${notFound}`);

        // Optionally write detailed results to a file
        const resultsOutput = results.map(r =>
            `${r.target}: ${r.found ? 'Found at index ' + r.index : 'Not found'}`
        ).join('\n');

        fs.writeFileSync('binary_search_results.txt', resultsOutput);

    } catch (error) {
        console.error('Error:', error.message);
        process.exit(1);
    }
}

// Run the main function
main();
